### ISSUE TITLE
Provide precompiled binaries for Mac's M1 ARM architecture

### ISSUE THREAD

**Issue by matthewmueller** [prisma/prisma#5245]
The new Macs now run on an ARM architecture. We should provide pre-compiled binaries for these machines.

We're currently waiting on: https://github.com/rust-lang/rust/issues/73908 before starting on this.

**Comment by leonardodino** [comment]
Hey there! 😄

I was able to successfully cross-compile `prisma-engine` binaries 🎉 
It was an uneventful and quite easy process. no errors on both compilation and run.
(although only tested sqlite adapter)

<details>
  <summary>Compiled on</summary>
> Mac mini (Late 2012)
> macOS Catalina 10.15.5 (19F101)

```
rustup show
Default host: x86_64-apple-darwin
rustup home:  /Users/leonardodino/.rustup

installed toolchains
--------------------

stable-aarch64-apple-darwin
stable-x86_64-apple-darwin (default)

installed targets for active toolchain
--------------------------------------

aarch64-apple-darwin
x86_64-apple-darwin

active toolchain
----------------

stable-x86_64-apple-darwin (default)
rustc 1.50.0 (cb75ad5db 2021-02-10)
```

</details>

(compiled on an old intel Mac Mini 2012, running Catalina + latest Xcode (12.4).
The binaries were tested on a sqlite project on a M1 MacBook Air, without rosetta installed.

(not sure if where to post this, as there's a related issue on the engines repo: https://github.com/prisma/prisma-engines/issues/1336)

**Comment by matthewmueller** [comment]
Just a note that Prisma does work with Rosetta emulation. We don't yet support native support for M1 though.

**Comment by Brooooooklyn** [comment]
Build Apple Silicon in Github Actions is easy, see: https://github.com/napi-rs/node-rs/blob/main/.github/workflows/ci.yaml#L474

**Comment by uncvrd** [comment]
@matthewmueller  hey! would you be able to clarify how to run prisma using rosetta emulation? I currently have the following Dockerfile but it fails to generate artifacts with the `npm install` prisma postinstall script. `prisma` is within my `package.json` which is installed upon build.

```
# ------------> Base Image
FROM node:14-alpine AS base

WORKDIR /workspace

# ------------> Development Image
FROM base AS development

COPY package*.json ./
COPY prisma ./prisma

RUN npm install

COPY . .

CMD ["npm", "run", "dev"]
```

Running this same `Dockerfile` on an intel mac works fine, but doesn't work on my new M1 device

Rosetta is working on my Mac as I use Rosetta version of Adobe Creative Cloud, but just not sure how to invoke rosetta for something like a `npm` package 🤔

**Comment by janpio** [comment]
@Brooooooklyn Seems the "official" instructions at https://github.com/shepmaster/rust/blob/silicon/silicon/README.md include `macosx11.0` sdk requirements, which are not available on `macos-latest` on Github Actions. How did you get around that?

I am getting the following error when using those instructions on the GH Actions `macos-latest` runners:
```
Run SDKROOT=$(xcrun -sdk macosx11.0 --show-sdk-path) \ MACOSX_DEPLOYMENT_TARGET=$(xcrun -sdk macosx11.0 --show-sdk-platform-version) \ cargo build  --target=aarch64-apple-darwin --release
xcodebuild: error: SDK "macosx11.0" cannot be located.
xcodebuild: error: SDK "macosx11.0" cannot be located.
xcrun: error: unable to lookup item 'PlatformVersion' in SDK 'macosx11.0'
xcodebuild: error: SDK "macosx11.0" cannot be located.
xcodebuild: error: SDK "macosx11.0" cannot be located.
xcrun: error: unable to lookup item 'Path' in SDK 'macosx11.0'
```

**Comment by janpio** [comment]
@uncvrd Optimally open a new issue about this and include all the information and the error message you are getting. It should "just work" usually, but might be different with Docker etc.

**Comment by Brooooooklyn** [comment]
@janpio see this: https://github.com/Brooooooklyn/hns/blob/main/.github/workflows/CI.yaml#L50-L55 . I actually found it from https://github.com/devongovett/dprint-node/blob/main/.github/workflows/tag-release.yml#L174-L179

**Comment by Brooooooklyn** [comment]
Don't know why some packages need this workaround https://github.com/Brooooooklyn/hns/blob/main/.github/workflows/CI.yaml#L50-L55. Most of my packages can directly build `aarch64-apple-darwin` target without any trick.

**Comment by janpio** [comment]
Thank you so much - your CI setup is amazing. 

I actually figured out that my problem was the GH runners now have MacOS 11.1 SDK installed, not the hardcoded 11.0 I copy pasted from https://github.com/shepmaster/rust/blob/silicon/silicon/README.md?rgh-link-date=2021-06-18T20%3A01%3A06Z - so fixing that to `11.1` did get me much further already.

**Comment by janpio** [comment]
Ok, spent a bit more this evening and it is working 🚀 

The built files on M1:
```
administrator@44528 m1BuildTest % file query-engine
query-engine: Mach-O 64-bit executable arm64
administrator@44528 m1BuildTest % file migration-engine
migration-engine: Mach-O 64-bit executable arm64
administrator@44528 m1BuildTest % file prisma-fmt
prisma-fmt: Mach-O 64-bit executable arm64
administrator@44528 m1BuildTest % file introspection-engine
introspection-engine: Mach-O 64-bit executable arm64
administrator@44528 m1BuildTest % file libquery_engine_napi.dylib
libquery_engine_napi.dylib: Mach-O 64-bit dynamically linked shared library arm64
```
The 4 binaries can all be executed 👍 

The same files on a normal Mac:
```
foo:binaries (5) foo$ ./introspection-engine 
-bash: ./introspection-engine: Bad CPU type in executable
foo:binaries (5) foo$ ./query-engine 
-bash: ./query-engine: Bad CPU type in executable
foo:binaries (5) foo$ ./migration-engine 
-bash: ./migration-engine: Bad CPU type in executable
foo:binaries (5) foo$ ./prisma-fmt 
-bash: ./prisma-fmt: Bad CPU type in executable
foo:binaries (5) foo$ ./libquery_engine_napi.dylib 
-bash: ./libquery_engine_napi.dylib: cannot execute binary file
```

https://github.com/prisma/prisma-engines/pull/2031

**Comment by janpio** [comment]
(Candidate for Client team @matthewmueller so @dpetrick has time to integrate the new Mac GH Actions build workflow into the engines build process)

**Comment by matthewmueller** [comment]
Questions:

- Is our build-step cross-compiling M1 from linux?

**Comment by janpio** [comment]
As written above you have to compile either on Apple Silicon directly, or on a normal macOS macOS machine where you then can compile for the `aarch64-apple-darwin` target architecture. You can see the full setup at https://github.com/prisma/prisma-engines/pull/2031/files

It is not (yet?) possible to cross compile Rust for Apple Silicon via Linux - at least nothing is documented in https://github.com/rust-lang/rust/issues/73908 and everyone (like @Brooooooklyn) is doing it the other way.

**Comment by janpio** [comment]
First engine build is available at https://binaries.prisma.sh/all_commits/c5c4a7b799133a6267e960c630fad57cc8a23799/darwin-arm64/libquery_engine.dylib.node.gz
- platform = `darwin-arm64`
- filename = `libquery_engine.dylib.node.gz`

**Comment by janpio** [comment]
This should now work with the latest `dev` version `2.27.0-dev.28`:

```
administrator@44528 engine24 % npx prisma -v
prisma               : 2.27.0-dev.28
@prisma/client       : Not found
Current platform     : darwin-arm64
Query Engine         : query-engine 0c2898954d761d1c92f304ff1b7917c601c2e3d8 (at node_modules/@prisma/engines/query-engine-darwin-arm64)
Migration Engine     : migration-engine-cli 0c2898954d761d1c92f304ff1b7917c601c2e3d8 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Introspection Engine : introspection-core 0c2898954d761d1c92f304ff1b7917c601c2e3d8 (at node_modules/@prisma/engines/introspection-engine-darwin-arm64)
Format Binary        : prisma-fmt 0c2898954d761d1c92f304ff1b7917c601c2e3d8 (at node_modules/@prisma/engines/prisma-fmt-darwin-arm64)
Default Engines Hash : 0c2898954d761d1c92f304ff1b7917c601c2e3d8
Studio               : 0.409.0
```

**Comment by Janushan** [comment]
Thanks this great worked for me!

**Comment by janpio** [comment]
This will be available for everyone with `2.27.0` next week, until then via `dev`.

We are also adding an end to end test to make sure this keeps working :D 
https://github.com/prisma/e2e-tests/pull/1798

**Comment by iddan** [comment]
I think the change broke installations of older versions of prisma in M1

**Comment by janpio** [comment]
Can you elaborate on that @iddan? The old CLI/Client should not know about this new binary `darwin-arm64` and just download and use the old one. Are you getting some error message? Then optimally open an issue with the error message and a bit of information on your project.

**Comment by iddan** [comment]
Yes, I'm having trouble to run tests in [prisma/nexus-prisma](https://github.com/prisma/nexus-prisma).

**Comment by janpio** [comment]
That would be really unexpected - please open a new issue and link to it here. Thanks.
